### PR TITLE
Fix BiDAF Model Serialization

### PR DESCRIPTION
**Pull request by nelson-liu**
@matt-gardner , here's the initial dump of files that we were just looking at.

### CODE REVIEW

**Comment by matt-gardner** [comment]
@nelson-liu, can you look this over?  Should be good to go.

**Comment by matt-gardner** [comment]
Nope, some tests still failing...  Let me fix this before you look at it.

**Comment by matt-gardner** [comment]
Ok, tests pass, it's good to look at.

**Comment by matt-gardner** [comment]
Ack, no, that was premature.  TF passes, theano fails...  Sorry for the spam.

File: deep_qa/layers/attention/matrix_attention.py [hunk 1]
```diff
@@ -87,10 +87,19 @@ def call(self, inputs, mask=None):
...
+        num_rows_1 = matrix_1_shape[1]
+        num_rows_2 = matrix_2_shape[1]
         tiled_matrix_1 = K.repeat_elements(K.expand_dims(matrix_1, dim=2), num_rows_2, axis=2)
         tiled_matrix_2 = K.repeat_elements(K.expand_dims(matrix_2, dim=1), num_rows_1, axis=1)
+
+        # We need to be able to access K.int_shape() in compute_similarity() below, but in theano,
+        # calling a backend function makes it so you can't use K.int_shape() anymore.  Setting
+        # tensor._keras_shape here fixes that.
+        # pylint: disable=protected-access
+        tiled_matrix_1._keras_shape = matrix_1_shape[:2] + (num_rows_2,) + matrix_1_shape[2:]
```
Comment (nelson-liu) [inline deep_qa/layers/attention/matrix_attention.py]: wow i wasnt aware this was a thing, nice!

**Comment by nelson-liu** [comment]
i cant approve my own PR, but this LGTM. feel free to merge if you don't have anything else to add

File: deep_qa/layers/attention/matrix_attention.py (hunk 1 above)
Comment (matt-gardner) [inline deep_qa/layers/attention/matrix_attention.py]: It's a total hack that we shouldn't rely on too extensively.  The better solution would be to fix all theano backend functions to set the `_keras_shape` on the returned tensor.  But I'm not in much of a mood to submit PRs to keras right now, and we're probably still in the PR freeze, anyway.

**Pull request merged**